### PR TITLE
BLS: add benchmarking for Affine form host functions

### PR DESCRIPTION
**Pull request by linh2931**
Benchmarking for Affine form host functions.

Did a comparison between Jacobian form and Affine form.

Jacobian form

```
function                                runs        average       minimum       maximum

bls:
bls_g1_add                              5,000       3,726 ns       3,546 ns      11,037 ns
bls_g1_add_mont                         5,000       3,423 ns       3,276 ns      11,356 ns
bls_g2_add                              5,000      10,195 ns       9,823 ns      22,700 ns
bls_g2_add_mont                         5,000       9,640 ns       9,219 ns      38,835 ns
bls_pairing 1 pair                      5,000   1,731,868 ns   1,682,579 ns   3,101,753 ns
bls_pairing_mont 1 pair                 5,000   1,736,806 ns   1,675,763 ns  12,641,838 ns
bls_pairing 3 pairs                     5,000   2,768,731 ns   2,653,448 ns   3,946,737 ns
bls_pairing_mont 3 pairs                5,000   2,757,585 ns   2,656,286 ns  16,153,583 ns
bls_g1_weighted_sum 1 point             5,000     185,576 ns     183,298 ns     207,651 ns
bls_g1_weighted_sum_mont 1 point        5,000     185,674 ns     183,339 ns     208,466 ns
bls_g1_weighted_sum 3 points            5,000     590,138 ns     579,728 ns     688,373 ns
bls_g1_weighted_sum_mont 3 points       5,000     606,476 ns     592,578 ns   8,403,153 ns
bls_g1_weighted_sum 5 points            5,000     735,405 ns     725,322 ns     845,926 ns
bls_g1_weighted_sum_mont 5 points       5,000     731,664 ns     721,941 ns     835,411 ns
bls_g2_weighted_sum 1 point             5,000     706,839 ns     685,898 ns   1,045,018 ns
bls_g2_weighted_sum_mont 1 point        5,000     743,762 ns     719,694 ns   1,011,417 ns
bls_g2_weighted_sum 3 points            5,000   2,301,903 ns   2,227,814 ns  16,066,772 ns
bls_g2_weighted_sum_mont 3 points       5,000   2,270,676 ns   2,205,879 ns   3,039,832 ns
bls_g2_weighted_sum 5 points            5,000   2,841,482 ns   2,759,238 ns  19,348,134 ns
bls_g2_weighted_sum_mont 5 points       5,000   2,829,128 ns   2,740,629 ns  14,205,987 ns
bls_g1_map                              5,000     301,257 ns     296,573 ns     442,878 ns
bls_g1_map_mont                         5,000     305,439 ns     301,122 ns     449,227 ns
bls_g2_map                              5,000     458,721 ns     448,555 ns     605,324 ns
bls_g2_map_mont                         5,000     453,444 ns     445,441 ns     602,810 ns
bls_fp_mod                              5,000         727 ns         625 ns       5,797 ns
bls_fp_mod_mont                         5,000         821 ns         687 ns       8,068 ns
bls_fp_mul                              5,000         681 ns         645 ns       4,119 ns
bls_fp_mul_mont                         5,000         588 ns         521 ns       6,536 ns
bls_fp_exp                              5,000      32,829 ns      31,899 ns      49,464 ns
bls_fp_exp_mont                         5,000      32,254 ns      31,151 ns      39,113 ns

```

Affine form

```
bls:
bls_g1_add                              5,000      76,699 ns      75,622 ns     123,329 ns
bls_g1_add_mont                         5,000      77,344 ns      75,954 ns     124,887 ns
bls_g2_add                              5,000      82,521 ns      81,196 ns     131,229 ns
bls_g2_add_mont                         5,000      82,184 ns      80,506 ns   1,454,766 ns
bls_pairing 1 pair                      5,000   1,638,763 ns   1,557,463 ns   3,569,653 ns
bls_pairing_mont 1 pair                 5,000   1,620,381 ns   1,545,184 ns   3,005,809 ns
bls_pairing 3 pairs                     5,000   2,321,344 ns   2,232,847 ns   3,740,820 ns
bls_pairing_mont 3 pairs                5,000   2,325,408 ns   2,249,310 ns   3,706,786 ns
bls_g1_weighted_sum 1 point             5,000     260,683 ns     258,196 ns     328,138 ns
bls_g1_weighted_sum_mont 1 point        5,000     256,633 ns     253,419 ns   1,086,645 ns
bls_g1_weighted_sum 3 points            5,000     670,546 ns     661,700 ns   2,038,831 ns
bls_g1_weighted_sum_mont 3 points       5,000     683,842 ns     675,641 ns   2,111,874 ns
bls_g1_weighted_sum 5 points            5,000     817,884 ns     808,866 ns   2,200,022 ns
bls_g1_weighted_sum_mont 5 points       5,000     807,745 ns     797,233 ns   2,215,336 ns
bls_g2_weighted_sum 1 point             5,000     818,862 ns     794,358 ns   2,368,434 ns
bls_g2_weighted_sum_mont 1 point        5,000     794,397 ns     770,952 ns   2,166,481 ns
bls_g2_weighted_sum 3 points            5,000   2,338,746 ns   2,285,345 ns   4,072,288 ns
bls_g2_weighted_sum_mont 3 points       5,000   2,382,895 ns   2,330,791 ns   4,336,254 ns
bls_g2_weighted_sum 5 points            5,000   2,924,081 ns   2,845,713 ns  22,833,113 ns
bls_g2_weighted_sum_mont 5 points       5,000   3,001,189 ns   2,914,939 ns  26,705,436 ns
bls_g1_map                              5,000     379,593 ns     372,223 ns     566,554 ns
bls_g1_map_mont                         5,000     375,586 ns     369,145 ns     563,225 ns
bls_g2_map                              5,000     531,899 ns     522,480 ns     722,943 ns
bls_g2_map_mont                         5,000     534,658 ns     520,874 ns     722,779 ns
bls_fp_mod                              5,000         794 ns         703 ns       9,219 ns
bls_fp_mod_mont                         5,000         862 ns         706 ns       6,692 ns
bls_fp_mul                              5,000         692 ns         621 ns       7,913 ns
bls_fp_mul_mont                         5,000         591 ns         540 ns       7,971 ns
bls_fp_exp                              5,000      32,397 ns      31,702 ns      43,478 ns
bls_fp_exp_mont                         5,000      32,535 ns      31,943 ns      44,466 ns
```

### CODE REVIEW

**Comment by greg7mdp** [comment]
Strange, it seems that providing the parameters in Montgomery form doesn't make a significant difference.

**Review by heifner — APPROVED** [review]

File: benchmark/bls.cpp [hunk 1]
```diff
@@ -122,11 +148,11 @@ bls12_381::g2 random_g2()
 }
 
 // bls_g1_add benchmarking
-void benchmark_bls_g1_add() {
+void benchmark_bls_g1_add_impl(std::string test_name, bool mont) {
```
Comment (greg7mdp) [inline benchmark/bls.cpp:151]: In this and other `_impl` instances, change `test_name` to  `const &` as it is not moved in the function.

File: benchmark/bls.cpp [hunk 2]
```diff
@@ -240,30 +243,57 @@ void benchmark_bls_g1_weighted_sum(std::string test_name, uint32_t num_points) {
...
+}
+
+// bls_g1_weighted_sum benchmarking with 5 input points
+void benchmark_bls_g1_weighted_sum_five_point() {
+   benchmark_bls_g1_weighted_sum_impl("bls_g1_weighted_sum 5 points", 5, false);
+}
+
+// bls_g1_weighted_sum benchmarking with 1 input point
+void benchmark_bls_g1_weighted_sum_mont_one_point() {
+   benchmark_bls_g1_weighted_sum_impl("bls_g1_weighted_sum_mont 1 point", 1, false);
```
[the remark this answers is not in the document]
Comment (linh2931) [reply benchmark/bls.cpp:277]: Thanks! You are correct. It should be `true`.

File: benchmark/bls.cpp [hunk 3]
```diff
@@ -107,6 +107,32 @@ std::array<uint64_t, 4> random_scalar()
...
+    std::mt19937_64 gen(rd());
+    std::uniform_int_distribution<uint64_t> dis;
+
+    return fp({
+        dis(gen) % 0xb9feffffffffaaab,
+        dis(gen) % 0x1eabfffeb153ffff,
+        dis(gen) % 0x6730d2a0f6b0f624,
+        dis(gen) % 0x64774b84f38512bf,
+        dis(gen) % 0x4b1ba7b6434bacd7,
+        dis(gen) % 0x1a0111ea397fe69a
```
[the remark this answers is not in the document]
Comment (linh2931) [reply benchmark/bls.cpp:123]: From the unit tests. Not sure what they are.

File: benchmark/bls.cpp [hunk 4]
```diff
@@ -122,11 +148,11 @@ bls12_381::g2 random_g2()
 }
 
 // bls_g1_add benchmarking
-void benchmark_bls_g1_add() {
+void benchmark_bls_g1_add_impl(std::string test_name, bool mont) {
    // prepare g1 operand in Jacobian LE format
    g1 p = random_g1();
    std::vector<char> buf(96);
-   p.toAffineBytesLE(std::span<uint8_t, 96>((uint8_t*)buf.data(), 96), true);
+   p.toAffineBytesLE(std::span<uint8_t, 96>((uint8_t*)buf.data(), 96), mont);
```
[the remark this answers is not in the document]
Comment (linh2931) [reply benchmark/bls.cpp:155]: `toBytesLE<96>(fromMont().d)` means input is Montgomery.

File: benchmark/bls.cpp [hunk 5]
```diff
@@ -135,19 +161,34 @@ void benchmark_bls_g1_add() {
 
    // set up bls_g1_add to be benchmarked
    interface_in_benchmark interface;
-   auto g1_add_func = [&]() {
-      interface.interface->bls_g1_add(op1, op1, result);
-   };
+   std::function<void()> benchmarked_func {};
+   if (mont) {
+      benchmarked_func = [&]() {
```
[the remark this answers is not in the document]
Comment (linh2931) [reply benchmark/bls.cpp:166]: Will need an API from BLS library. It is probably OK not to add an assert as this is a benchmarking program. When input was wrong, the result would be obvious (very small nubmers)

File: benchmark/bls.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (greg7mdp) [reply benchmark/bls.cpp:155]: Yes, I agree. If the input is Montgomery, shouldn't we call `fromMont()`?
In this case, even though the input is Montgomery, we don't.

**Comment by yarkinwho** [comment]
All _mont function removed now. Benchmarks for them should be removed.

File: benchmark/bls.cpp [hunk 6]
```diff
@@ -395,27 +503,144 @@ void benchmark_bls_fp_mod() {
...
+   benchmark_bls_fp_mul_impl("bls_fp_mul", false);
+}
+
+void benchmark_bls_fp_mul_mont() {
+   benchmark_bls_fp_mul_impl("bls_fp_mul_mont", true);
+}
+
+void benchmark_bls_fp_exp_impl(const std::string& test_name, bool mont) {
+   // prepare base
+   std::vector<char> base_buf(48);
```
[the remark this answers is not in the document]
Comment (linh2931) [reply benchmark/bls.cpp]: Thanks! Done.

**Comment by linh2931** [comment]
> All _mont function removed now. Benchmarks for them should be removed.

Thanks for the reminder. Removed.

**Review by greg7mdp — APPROVED** [review]

**Review by yarkinwho — APPROVED** [review]

**Pull request merged**